astropy/utils/data.py

```diff
@@ -232,6 +233,7 @@ def get_readable_fileobj(name_or_obj, encoding=None, cache=False,
         try:
             import bz2
         except ImportError:
+            fileobj.close()
```

Comment:
Do not apologise too soon - apparently the finally does not catch it as it's not in the same scope, anyway after an `ImportError` I got open-file errors here. Changed it to your suggestion.